ignite/engine/engine.py

```diff
@@ -123,28 +123,48 @@ def compute_mean_std(engine, batch):
     _state_dict_one_of_opt_keys = ("iteration", "epoch")
 
     def __init__(self, process_function: Callable):
-        self._event_handlers = defaultdict(list)  # type: Dict[Any, List]
+        super(Engine, self).__init__()
         self.logger = logging.getLogger(__name__ + "." + self.__class__.__name__)
         self._process_function = process_function
-        self.last_event_name = None  # type: Optional[Events]
         self.should_terminate = False
         self.should_terminate_single_epoch = False
-        self.state = State()
         self._state_dict_user_keys = []  # type: List[str]
-        self._allowed_events = []  # type: List[EventEnum]
 
         self._dataloader_iter = None  # type: Optional[Iterator[Any]]
         self._init_iter = []  # type: List[int]
 
-        self.register_events(*Events)
+        self.register_events(*Events, attr_to_events=State.attr_to_events)
+        self._state = State(engine=self)
 
         if self._process_function is None:
             raise ValueError("Engine must be given a processing function in order to run.")
 
         _check_signature(process_function, "process_function", self, None)
 
+    @property
+    def state(self) -> State:
+        return self._state  # type: ignore
+
+    @state.setter
+    def state(self, new_state: State) -> None:
+        for new_events in new_state._attr_to_events.values():
+            if new_events not in self._state._attr_to_events.values():
+                raise ValueError("The new state must not contain any new unseen events.")
```

Comment:
How about using sets for that ? 
```python
a = {1, 2, 3}
b = {2, 3}
a - b
```

Reply:
Values here are list of list, which are not hashable, we can't convert them into sets.